src/metabase/db/data_source.clj

```diff
@@ -32,36 +32,45 @@
-    ;; (I think this is also how some places like Heroku ship them in order to make our lives hard) So strip those out
-    ;; with the absolute minimum of parsing we can get away with and then pass them in separately -- see #14678 and
-    ;; #20121
-    ;;
-    ;; NOTE: if password is URL-encoded this isn't going to work, since we're not URL-decoding it. I don't think that's
-    ;; a problem we really have to worry about, and at any rate we have never supported it. We did URL-decode things at
-    ;; one point, but that was only because [[clojure.java.jdbc]] tries to parse connection strings itself if you let it
-    ;; -- see #14836. We never let it see connection strings anymore, so that shouldn't be a problem. At any rate #20122
-    ;; would probably solve most people's problems if their password contains special characters.
-    (if-let [[_ subprotocol user password more] (re-find #"^jdbc:((?:postgresql)|(?:mysql))://([^:@]+)(?::([^@:]+))?@(.+$)" s)]
-      (->DataSource (str "jdbc:" subprotocol "://" more)
-                    (pool/map->properties
-                     (merge {:user user}
-                            (when (seq password)
-                              {:password password}))))
-      (->DataSource s nil))))
+  (^javax.sql.DataSource [s]
+   (raw-connection-string->DataSource s nil nil))
+
+  (^javax.sql.DataSource [s username password]
+   {:pre [(string? s)]}
+   ;; normalize the protocol in case someone is trying to trip us up. Heroku is known for this and passes stuff in
+   ;; like `postgres:...` to screw with us.
+   (let [s (cond-> s
+             (str/starts-with? s "postgres:")   (str/replace-first #"^postgres:" "postgresql:")
+             (not (str/starts-with? s "jdbc:")) (str/replace-first #"^" "jdbc:"))]
+     ;; Even tho they're invalid we need to handle strings like `postgres://user:password@host:port` for legacy reasons.
+     ;; (I think this is also how some places like Heroku ship them in order to make our lives hard) So strip those out
+     ;; with the absolute minimum of parsing we can get away with and then pass them in separately -- see #14678 and
+     ;; #20121
+     ;;
+     ;; NOTE: if password is URL-encoded this isn't going to work, since we're not URL-decoding it. I don't think that's
+     ;; a problem we really have to worry about, and at any rate we have never supported it. We did URL-decode things at
+     ;; one point, but that was only because [[clojure.java.jdbc]] tries to parse connection strings itself if you let it
+     ;; -- see #14836. We never let it see connection strings anymore, so that shouldn't be a problem. At any rate #20122
+     ;; would probably solve most people's problems if their password contains special characters.
+     (let [[s m] (if-let [[_ subprotocol user password more] (re-find #"^jdbc:((?:postgresql)|(?:mysql))://([^:@]+)(?::([^@:]+))?@(.+$)" s)]
```

Comment:
just noticed this is a redundant let

Reply:
Not really redundant since that form returns a `string-map` tuple `[s m]` in different ways based on the `if-let` inside it. `let` inside `let` is a little out of the ordinary but it's not redundant

Reply:
not that `let`. It's the let that's above the comment. `(let [s (cond-> ...)] (let [[s m] ...)`

Reply:
fixed